plugins/outputs/firehose/firehose.go

```diff
@@ -0,0 +1,268 @@
+  ## 6) EC2 Instance Profile
+  #access_key = ""
+  #secret_key = ""
+  #token = ""
+  #role_arn = ""
+  #profile = ""
+  #shared_credential_file = ""
+
+  ## Firehose StreamName must exist prior to starting telegraf.
+  delivery_stream_name = "FirehoseName"
+
+  ## The maximum number of times to attempt resubmitting a metric batch
+  max_submit_attempts = 10
+
+  ## Whether or not to enable gzip compression. Compression rates are usually
+  ## as good as 90% space savings. However, the destination must decompress the data 
+  ## before it goes to influx.
+  ## The good news here, gzip it designed to be concat'ed together. We will compress 
+  ## all metrics received during a telegrate write and send them to firehose. Firehose will
+  ## then concat all this data together- but a single decompress will get all lines back in
+  ## a single pass.  All metric lines will have a newline character seperating them.
+  enable_gzip_compression = false
+
+  ## Data format to output.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  data_format = "influx"
+
+  ## debug will show upstream aws messages.
+  debug = false
+`
+
+func (f *FirehoseOutput) SampleConfig() string {
+	return sampleConfig
+}
+
+func (f *FirehoseOutput) Description() string {
+	return "Configuration for the AWS Firehose output."
+}
+
+func (f *FirehoseOutput) Connect() error {
+	// We attempt first to create a session to Firehose using an IAMS role, if that fails it will fall through to using
+	// environment variables, and then Shared Credentials.
+	if f.Debug {
+		log.Printf("E! firehose: Building a session for connection to Firehose in %+v", f.Region)
+	}
+
+	credentialConfig := &internalaws.CredentialConfig{
+		Region:    f.Region,
+		AccessKey: f.AccessKey,
+		SecretKey: f.SecretKey,
+		RoleARN:   f.RoleARN,
+		Profile:   f.Profile,
+		Filename:  f.Filename,
+		Token:     f.Token,
+	}
+	configProvider := credentialConfig.Credentials()
+
+	// we simply create the skeleton here. AWS doesn't attempt
+	// any connection yet, so we don't know if an error will happen.
+	svc := firehose.New(configProvider)
+	f.svc = svc
+	return nil
+}
+
+func (f *FirehoseOutput) Close() error {
+	return nil
+}
+
+func (f *FirehoseOutput) SetSerializer(serializer serializers.Serializer) {
+	f.serializer = serializer
+}
+
+// writeToFirehose uses the AWS GO SDK to write batched records to firehose and
+// queue the failed writes to the errorBuffer of f
+func writeToFirehose(f *FirehoseOutput, b [][]byte, submitAttemptCount int64) {
+	start := time.Now()
+	dataSize := 0 // track the size of the record in bytes
+	var values []byte
+
+	// take metrics and flatten into a record
+
+	if f.EnableGzipCompression == true {
+		var buf bytes.Buffer
+		zw := gzip.NewWriter(&buf)
+
+		for _, line := range b {
+			_, err := zw.Write(line)
+			if err != nil {
+				log.Fatal(err)
```

Comment:
Don't `log.Fatal`, for errors that shouldn't stop processing you can `log.Printf("E! `, or if it should fail the entire batch you should return the error.